### PR TITLE
virtual wxListCtrl checkboxes

### PR DESCRIPTION
**Pull request by MaartenBent**
Enable checkboxes in a virtual `wxListCtrl` (on `wxMSW`).
See https://trac.wxwidgets.org/ticket/18393

### CODE REVIEW

File: samples/listctrl/listtest.cpp [hunk 1]
```diff
@@ -1204,13 +1205,21 @@ void MyListCtrl::OnChecked(wxListEvent& event)
 {
     LogEvent(event, "OnChecked");
 
+    if ( IsVirtual() ) {
```
Comment (vadz) [inline samples/listctrl/listtest.cpp]: Usual formatting nitpick (here and below):
```suggestion
    if ( IsVirtual() )
    {
```

File: samples/listctrl/listtest.h [hunk 2]
```diff
@@ -83,11 +86,15 @@ class MyListCtrl: public wxListCtrl
...
 
     virtual wxString OnGetItemText(long item, long column) const wxOVERRIDE;
+    virtual bool OnGetItemIsChecked(long item) const wxOVERRIDE;
     virtual int OnGetItemColumnImage(long item, long column) const wxOVERRIDE;
     virtual wxItemAttr *OnGetItemAttr(long item) const wxOVERRIDE;
 
     long m_updated;
 
+    // checked boxes in virtual list
+    WX_DECLARE_HASH_MAP(long, bool, wxIntegerHash, wxIntegerEqual, wxLongBoolMap);
```
Comment (vadz) [inline samples/listctrl/listtest.h]: Could also use `wxSelectionStore` here: in spite of its name, it should be equally suitable for storing checked state as selected one.

File: src/generic/listctrl.cpp [hunk 3]
```diff
@@ -5542,6 +5542,15 @@ wxString wxGenericListCtrl::OnGetItemText(long WXUNUSED(item), long WXUNUSED(col
     return wxEmptyString;
 }
 
+bool wxGenericListCtrl::OnGetItemIsChecked(long WXUNUSED(item)) const
+{
+    // this is a pure virtual function, in fact - which is not really pure
+    // because the controls which are not virtual don't need to implement it
+    wxFAIL_MSG(wxT("wxGenericListCtrl::OnGetItemIsChecked not supposed to be called"));
```
Comment (vadz) [inline src/generic/listctrl.cpp]:
```suggestion
    wxFAIL_MSG(wxS("wxGenericListCtrl::OnGetItemIsChecked not supposed to be called"));
```
Or just remove it entirely. But there is really no good reason to use `wxT` in the new code.

File: include/wx/generic/listctrl.h [hunk 4]
```diff
@@ -216,6 +216,9 @@ class WXDLLIMPEXP_CORE wxGenericListCtrl: public wxNavigationEnabled<wxListCtrlB
     // return the text for the given column of the given item
     virtual wxString OnGetItemText(long item, long column) const;
 
+    // return whether the given item is checked
+    virtual bool OnGetItemIsChecked(long item) const;
```
Comment (vadz) [inline include/wx/generic/listctrl.h]: This is consistent with the other methods, but seems a bit awkward to me. I wonder if `OnItemIsChecked()` could be better while still retaining a degree of consistency?

But I'm leaving the final choice of the name to you here, this is just a suggestion.

**Review by vadz — COMMENTED** [review]
First of all, thanks for implementing it, it definitely makes sense to support this.

In addition to the small remarks below, I wonder if there is anything preventing us from having a single function stub in `wxListCtrlBase` instead of duplicating it in all ports?

**Comment by MaartenBent** [comment]
I applied your suggestions. I prefer keeping `OnGetItemIsChecked` so it matches the other virtual functions.

**Comment by vadz** [comment]
Will merge in a moment, thanks again!

**Pull request merged**